### PR TITLE
Implemented factory - proxy pattern

### PR DESCRIPTION
**Pull request by amiromayer**

### CODE REVIEW

File: contracts/contracts/Factory.sol [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+    }
+  }
+
+    
+    // Deploy new proxy contract
+    function deployProxy(address payable _sender) 
+    external 
+    returns (address payable) 
+    {
+        // address payable proxy = createClone(masterCopy, uint(_sender));
```
Comment (Dobrokhvalov) [inline contracts/contracts/Factory.sol]: remove comments

File: contracts/src/utils.js [hunk 2]
```diff
@@ -0,0 +1,28 @@
+import { utils } from 'ethers'
```
Comment (Dobrokhvalov) [inline contracts/src/utils.js]: change `src` folder name to different name, e.g. `lib` or `utils`

File: contracts/contracts/Factory.sol [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+    {
+        code = _code;
+        implementation = _implementation;
+    }
+
+    function getSalt(address payable sender) public pure returns (bytes32) {
+        return keccak256(abi.encodePacked(sender));
+    }
+
+    function getCreate2Address
```
Comment (Dobrokhvalov) [inline contracts/contracts/Factory.sol]: getCreate2Address -> precomputeAddress

File: contracts/contracts/Factory.sol [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+        )
+      )
+    );
+
+    // determine if any contract code already exists at the computed address.
+    assembly { // solhint-disable-line
+      existingContractSize := extcodesize(deploymentAddress)
+    }
+
+    // if so, return null address to signify failure. (detect selfdestructed?)
```
Comment (Dobrokhvalov) [inline contracts/contracts/Factory.sol]: why do we need it here? How is that function is supposed to be used?

File: contracts/contracts/Factory.sol [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+
+        address payable proxy;
+
+        bytes memory _code = code;
+        bytes32 salt = getSalt(_sender);
+        assembly {
+            proxy := create2(0, add(_code, 0x20), mload(_code), salt)
+            if iszero(extcodesize(proxy)) {revert(0, 0)}
+        }
+        contractsDeployed++;
```
Comment (Dobrokhvalov) [inline contracts/contracts/Factory.sol]: why do we need counter and mapping here? How we gonna use it?

File: contracts/contracts/Factory.sol (hunk 4 above)
Comment (amiromayer) [reply contracts/contracts/Factory.sol]: Its from here: https://github.com/0age/Pr000xy/blob/c293074509c0f0d5fdf54c451ab39b2e00f48b7a/contracts/Create2Factory.sol#L68

File: contracts/contracts/Factory.sol (hunk 5 above)
Comment (amiromayer) [reply contracts/contracts/Factory.sol]: Needed them for testing purposes, now removed to avoid extra code

**Comment by amiromayer** [comment]
@Dobrokhvalov Commited

**Comment by Dobrokhvalov** [comment]
Closing the PR, as the chosen approach is 2-3 times more expensive than EIP-1167 ( eips.ethereum.org/EIPS/eip-1167 )

**Pull request closed without merging**